### PR TITLE
feat: LoRA SFT support for DTensorV2 path

### PR DESCRIPTION
**Pull request by samodi-nv**
# Issues
Addresses #833

# Test with thinking machine config
[Reproduce Recipe](https://github.com/NVIDIA-NeMo/RL/pull/1556/files#diff-27ecb93dbbdae7823c2f55c526d32da799b472f88a9ea8f40f6e4b3be26d9b99) : Tulu3 dataset is not supported yet. So this test can not be used for now. Re-enable this test once PR https://github.com/NVIDIA-NeMo/RL/pull/1506 is merged.

Or you can cherry pick [tulu3 dataset](https://github.com/NVIDIA-NeMo/RL/pull/1506/files#diff-e9defdd07971a08059e729f9340f0316ac0d18e900a860b94ba560f2df3ee79a) for your local branch and modified [corresponding nemo_rl/data/datasets/response_datasets/__init__.py](https://github.com/NVIDIA-NeMo/RL/pull/1506/files#diff-15be4dfa015665ac9aba22cabddab4c2fb7a9961e54c28172578dec325b502d9R99) as well.

image

# Description
This PR is a a work in progress to add LoRA support for the DTensor path. 

Current status
- [x] Add functionality to SFT path
- [x] Verify LoRA accuracy compared to [Thinking Machines blog](https://thinkingmachines.ai/blog/lora/)

# Notes
1. Support SFT + Lora. The result aligned with [Thinking Machines blog](https://thinkingmachines.ai/blog/lora/). 
2. The previous grad spike was due to a bug in the automodel initialization method. [Modifications to the automodel ](https://github.com/NVIDIA-NeMo/Automodel/commit/2d20e33a19d5e53a271b1403b507475e68ad14dc) have been merged into the main branch of the automodel. However, because our submodule is currently using a version significantly different from the main branch, [a patch](https://github.com/NVIDIA-NeMo/RL/pull/1556/files#diff-6bb1bc4a991f4db9244a9bd83f5e8aba519af97f3e82e58dc1eb4fbfbe2c0700R103) has been applied.
3. Our current commit usage for the automodel submodule is somewhat outdated. @RayenTian will later create a new dedicated branch for nemorl within the automodel repository and [dump the changes to this branch](https://github.com/NVIDIA-NeMo/RL/issues/1586). Once complete, [this patch](https://github.com/NVIDIA-NeMo/RL/pull/1556/files#diff-6bb1bc4a991f4db9244a9bd83f5e8aba519af97f3e82e58dc1eb4fbfbe2c0700R103) can be deleted.

## Summary by CodeRabbit

* **New Features**
  * Added LoRA (Low-Rank Adaptation) configuration support for parameter-efficient fine-tuning in supervised fine-tuning workflows, including customizable settings for module targeting, dimensionality, and dropout.
  * LoRA weights are now properly handled during checkpoint saving and loading.

* **Tests**
  * Added functional and unit tests for LoRA-enabled training and checkpoint management.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: nemo_rl/models/policy/__init__.py [hunk 1]
```diff
@@ -21,6 +21,19 @@ class DTensorConfigDisabled(TypedDict):
...
+    target_modules: NotRequired[list[str]]
+    exclude_modules: NotRequired[list[str]]
+    match_all_linear: NotRequired[bool]
+    dim: NotRequired[int]
+    alpha: NotRequired[int]
+    dropout: NotRequired[float]
+    dropout_position: NotRequired[Literal["pre", "post"]]
+    lora_A_init: NotRequired[str]
+    use_triton: NotRequired[bool]
+
```
[the remark this answers is not in the document]
Comment (joyang-nv) [reply nemo_rl/models/policy/__init__.py:35]: I think we need to fix this.

File: tests/functional/test_automodel_lora_sft.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    sft.val_period=3 \
+    policy.dtensor_cfg.lora.enabled=true \
+    logger.tensorboard_enabled=true \
+    logger.log_dir=$LOG_DIR \
+    logger.wandb_enabled=false \
+    logger.monitor_gpus=true \
+    checkpointing.enabled=true \
+    checkpointing.save_period=3 \
+    checkpointing.checkpoint_dir=/tmp/lora_sft_checkpoints \
+    $@ \
```
[the remark this answers is not in the document]
Comment (joyang-nv) [reply tests/functional/test_automodel_lora_sft.sh]: Oh, I think this is a good catch!

File: tests/unit/utils/test_automodel_checkpoint.py [hunk 3]
```diff
@@ -418,3 +473,83 @@ def test_save_and_load_model_and_optimizer(self, mock_experiment):
...
+        # lora_B: (out_features, dim) = (1, 2)
+        assert lora_state_dict["layers.3.lora_A.weight"].shape == (2, 4), (
+            f"Expected layers.3.lora_A.weight shape (2, 4), got {lora_state_dict['layers.3.lora_A.weight'].shape}"
+        )
+        assert lora_state_dict["layers.3.lora_B.weight"].shape == (1, 2), (
+            f"Expected layers.3.lora_B.weight shape (1, 2), got {lora_state_dict['layers.3.lora_B.weight'].shape}"
+        )
+
+        initial_distribute = torch.distributed.is_initialized()
+        print(f"Initial distribute: {initial_distribute}")
```
[the remark this answers is not in the document]
Comment (joyang-nv) [reply tests/unit/utils/test_automodel_checkpoint.py]: Only P1. :)

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 4]
```diff
@@ -234,6 +262,9 @@ def __init__(
                 torch_dtype=str(model_config.torch_dtype),
             )
 
+            if self.peft_config is not None:
+                apply_lora_to_linear_modules(model, self.peft_config)
```
Comment (hemildesai) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:269]: Looks like this function is called twice (on this line and line 291). Is that expected?

File: nemo_rl/models/policy/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (terrykong) [reply nemo_rl/models/policy/__init__.py:35]: could you make all of these required if they are required when `enabled=True`? you can get the type checker to respect notrequired if false and required if true by doing something like this:

https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/models/policy/__init__.py#L191

File: nemo_rl/models/policy/__init__.py [hunk 5]
```diff
@@ -21,6 +21,19 @@ class DTensorConfigDisabled(TypedDict):
...
+    enabled: bool
+    target_modules: NotRequired[list[str]]
+    exclude_modules: NotRequired[list[str]]
+    match_all_linear: NotRequired[bool]
+    dim: NotRequired[int]
+    alpha: NotRequired[int]
+    dropout: NotRequired[float]
+    dropout_position: NotRequired[Literal["pre", "post"]]
+    lora_A_init: NotRequired[str]
+    use_triton: NotRequired[bool]
```
Comment (terrykong) [inline nemo_rl/models/policy/__init__.py]: could you please help document these flags?

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 6]
```diff
@@ -94,6 +100,15 @@
 from nemo_rl.utils.packed_tensor import packed_broadcast_producer
 
 
+# TODO: @ruit remove this once the bump Automodel to 2d20e33a19d5e53a271b1403b507475e68ad14dc (https://github.com/NVIDIA-NeMo/RL/issues/1586)
+def _patched_init_lora_weights(self, init_method: str):
```
Comment (terrykong) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:103]: i can see this function silently getting ignored. could you add a unit test that does a check and will fail when it's okay to remove? something in spirit to this https://github.com/NVIDIA-NeMo/RL/blob/40e704072b78eb77d755afbc16c30c99ca7b612e/tests/unit/models/huggingface/test_smolvlm_embeddings_bug.py#L172

File: nemo_rl/models/policy/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (terrykong) [reply nemo_rl/models/policy/__init__.py:35]: i see later on there's this comment about the lora_dtype
image

but will that override the value chosen here. +1 on resolving this before merging

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 4 above)
Comment (RayenTian) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:269]: Yes. We need call it twice. At the first time, lora is initiated with weight and is applied to `model` to get model `full_state_dict`. For the second time, lora is applied to `self.model` without weight. Later we will assign the values of `full_state_dict` to self.model. We must call this twice to ensure their structures are consistent.

File: tests/functional/test_automodel_lora_sft.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (RayenTian) [reply tests/functional/test_automodel_lora_sft.sh]: Done.

File: tests/unit/utils/test_automodel_checkpoint.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RayenTian) [reply tests/unit/utils/test_automodel_checkpoint.py]: Done.

File: nemo_rl/models/policy/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (RayenTian) [reply nemo_rl/models/policy/__init__.py:35]: The lora type must same as the model type, so I removed this attribute in lora config to avoid users setting it wrong.

File: nemo_rl/models/policy/__init__.py (hunk 5 above)
Comment (RayenTian) [reply nemo_rl/models/policy/__init__.py]: Done. Added it in docs/guides/sft.md.

**Comment by RayenTian** [comment]
> Can you also add a test case similar to `3rdparty/Automodel-workspace/Automodel/tests/unit_tests/_peft/test_lora.py::test_forward_output_consistency()`?

Done. I’ve added [some tests here](https://github.com/NVIDIA-NeMo/RL/pull/1556/files#diff-6cc6e6db5d670f4becaca5ce749c5cd3781f6587fef6802f976206c26a30de85R138). That said, I’m not certain if these tests are necessary—they focus on testing AutoModel’s LoRA functionality rather than being related to NeMo-RL itself.

File: nemo_rl/models/policy/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (RayenTian) [reply nemo_rl/models/policy/__init__.py:35]: > could you make all of these required if they are required when `enabled=True`? you can get the type checker to respect notrequired if false and required if true by doing something like this:
> 
> https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/models/policy/__init__.py#L191

Done.

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 6 above)
Comment (RayenTian) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:103]: Done.  Add it at [tests/unit/models/dtensor/test_lora.py](https://github.com/NVIDIA-NeMo/RL/pull/1556/files#diff-6cc6e6db5d670f4becaca5ce749c5cd3781f6587fef6802f976206c26a30de85R41)

File: docs/guides/sft.md [hunk 7]
```diff
@@ -161,3 +161,45 @@ As long as your custom dataset has the `formatted_ds` and `task_spec` attributes
...
+## LoRA Configuration
+
+NeMo RL supports LoRA (Low-Rank Adaptation) for parameter-efficient fine-tuning. LoRA reduces trainable parameters by using low-rank matrices for weight updates while keeping the base model frozen.
+
+### Configuration Parameters
+
+The LoRA configuration is specified under the `policy.lora_cfg` section:
+
+policy:
+  lora_cfg:
```
Comment (terrykong) [inline docs/guides/sft.md]: So here's the mbridge one:
https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/b3571fbc96d6e2149215734a1e9e5e9ef74d9c1b/src/megatron/bridge/peft/lora.py#L56-L77

and here's the automodel one

https://github.com/NVIDIA-NeMo/Automodel/blob/5dfc3a69cb08cd11e2cf4559ffda60edd894c35b/nemo_automodel/components/_peft/lora.py#L37

they aren't exactly the same. rather than figure out the common feature set, i think for now it's maybe better to keep them disjoint within their respective backend config. in other words put this inside `dtensor_cfg`

In terms of config overlap, we can just use
```yaml
policy:
  megatron_cfg:
    lora_cfg.dim: ${policy.dtensor_cfg.lora_cfg.dim}
```
to tie the common ones together

File: examples/configs/recipes/llm/sft-llama3.2-1b-1n8g-fsdp2tp1.v3.yaml
Comment (terrykong) [inline examples/configs/recipes/llm/sft-llama3.2-1b-1n8g-fsdp2tp1.v3.yaml:1]: so a convention we follow is if you change a recipe, you have to add a `v$N` modifier so that it's clear to users that this version of the recipe is not comparable with the previous (just in case we're debugging a convergence/perf issue)

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 8]
```diff
@@ -233,6 +265,9 @@ def __init__(
                 torch_dtype=str(model_config.torch_dtype),
             )
 
+            if self.peft_config is not None:
```
Comment (terrykong) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: nit: can this be made consistent with the checks below

```suggestion
            if self.lora_enabled:
```

File: tests/unit/utils/test_automodel_checkpoint.py [hunk 9]
```diff
@@ -66,6 +74,53 @@ def mock_optimizer():
...
+@pytest.fixture
+def mock_distributed():
+    """Mock torch.distributed calls for non-distributed tests."""
+    with (
+        patch("torch.distributed.is_initialized", return_value=False),
+        patch("torch.distributed.get_rank", return_value=0),
+    ):
+        yield
+    torch.distributed.destroy_process_group()
+
```
[the remark this answers is not in the document]
Comment (terrykong) [reply tests/unit/utils/test_automodel_checkpoint.py]: this comment seems correct @RayenTian

[the remark this answers is not in the document]
Comment (RayenTian) [reply tests/unit/utils/test_automodel_checkpoint.py]: I checked that `mock_distributed` never be used actually. Removed it in latest commit.

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 8 above)
Comment (RayenTian) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: Done.

File: docs/guides/sft.md (hunk 7 above)
Comment (RayenTian) [reply docs/guides/sft.md]: Done.

File: examples/configs/recipes/llm/sft-llama3.2-1b-1n8g-fsdp2tp1.v3.yaml
Comment (RayenTian) [reply examples/configs/recipes/llm/sft-llama3.2-1b-1n8g-fsdp2tp1.v3.yaml:1]: Add a new config to test lora. Then this file has no change.

**Review by joyang-nv — APPROVED** [review]
